xarray/backends/api.py

```diff
@@ -79,6 +83,42 @@ def check_name(name):
         check_name(k)
 
 
+def _validate_attrs(dataset):
+    """`attrs` must have a string key and a value which is either: a number
+    a string, an ndarray or a list/tuple of numbers/strings.
+    """
+    def check_attr(name, value):
+        if isinstance(name, basestring):
+            if not name:
+                raise ValueError('Invalid name for attr: string must be length '
+                                 '1 or greater for serialization to netCDF '
+                                 'files')
+        else:
+            raise TypeError('Invalid name for attr: must be a string for '
+                            'serialization to netCDF files')
+
+        if not isinstance(value, basestring) and not isinstance(value, Number) and \
+           not isinstance(value, np.ndarray) and not isinstance(value, list) and \
+           not isinstance(value, tuple):
+                raise TypeError('Invalid value for attr: must be a number '
+                                'string, ndarray or a list/tuple of numbers/strings '
+                                'for serialization to netCDF files')
+
+    # Check attrs on the dataset itself
+    for k, v in dataset.attrs.items():
+        check_attr(k, v)
+
+    # Check attrs on each variable within the dataset
+    for var_name in dataset.data_vars:
```

Comment:
Use `dataset.variables`, which is a dict of [`xarray.Variable` objects](http://xarray.pydata.org/en/stable/internals.html#variable-objects)